src/sentry/plugins/providers/integration_repository.py

```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
 
     def dispatch(self, request, organization, **kwargs):
         try:
+            # get data via API (so we know it's up to date)
             config = self.get_repository_data(organization, request.DATA)
+
+        except Exception as e:
+            return self.handle_api_error(e)
+
+        else:
+            # check to see if we already have this repository, and update any
+            # data if applicable
+            repo = Repository.objects.get(
+                organization_id=organization.id,
+                provider=self.id,
+                external_id=config.get('external_id')
+            )
+            if repo is not None:
+                updated = self.update_repository_data(repo, config)
+                if updated:
+                    return Response(serialize(repo, request.user), status=200)
+                return Response({
+                    'errors': {
+                        '__all__': 'A repository with that name already exists'
+                    }
+                },
+                    status=400,
+                )
+
+        # at this point we know it's a new repo
+        try:
             result = self.build_repository_config(
                 organization=organization,
                 data=config,
             )
         except Exception as e:
             return self.handle_api_error(e)
 
+        # TODO do we need all the try-except business here anymore, or are we
```

Comment:
We're catching an `IntegrityError`, which I believe `.create()` only throws in the case of duplicate primary keys/violation of unique constraints. (Correct me if I'm wrong about that.) If that's true, I believe those cases should already have been handled above.

Reply:
I agree. I'm not convinced those two should be separate. Could you give me a sense as to you handled it separately as opposed to a `create_or_update`?

Reply:
It's partly because the case of creation was already handled separately from the case of it already existing (in which case we had to go undo the side effects of attempting to create the repo, namely deleting the webhook which got created).

Regardless, I think we need to keep it separate for two reasons:

1) We really have three cases, not two:

    a) the repo doesn't exist - create it
    b) the repo does exist but some of the data is wrong - update it
    c) the repo exists and none of the data is different - send back an error

    Case b) is the one I added. Switching to a `create_or_update` would force b) and c) to be the same case, so we'd have to either always send back an error, or never send back an error (the latter being arguably better, but also a change from current behavior).

2) In any case, even if we wanted to do that, unfortunately the "update" part of `create_or_update` isn't as simple as one might like - different information needs to be updated depending on what integration we're talking about (specifically in the `config` column), so I'm not clear how we'd consolidate the behavior into this central location. (That's why I forced each integration to implement its own `update_repository_data` the same way each one has its own `get_repository_data` and `build_repository_config`.)

    The alternative would be to split creation out into the different integrations also... or to call `build_repository_config` _before_ the possible `create_or_update` (so we'd have the correct `config` column value), but then we're exactly back where we started: `build_repository_config` creates a webhook, and we don't want to do that if all we're doing is updating, so we'd have to go and delete it, just like we're trying to get away from.

I do think it would be simpler to just do it all here in `integration_repository.py` and/or with a `create_or_update`, but I just couldn't figure out how to do that given the above.

Reply:
Ok I see. Let's just get the repo. Like you're doing here. I think that'll be the clearer answer. I had thought of `get_or_create` as well. But ultimately I'm not sure we save quite that much and probably lose in terms of readability.